### DISCUSSION TITLE
Colored cell

### DISCUSSION THREAD

**Discussion by hoclun-rigsep** [saulpw/visidata#1088] [Q&A]
Hello,

I have an Excel sheet in which cell color has semantic value. Any way to access this information with VisiData? I tried to dig down to the PyObject level to see if I could find cell color somewhere in there, but I have not been successful.

Thank you!
~hoclun-rigsep

**Comment by saulpw** [comment]
Hi @hoclun-rigsep, there isn't currently any way.  If you find a way for the `openpyxl` to expose that information, then we can figure out a way to get it to you.

**Reply by hoclun-rigsep** [reply]
This is what I have so far:

```diff
diff --git a/visidata/loaders/xlsx.py b/visidata/loaders/xlsx.py
index f466cf92..ff05c0e7 100644
--- a/visidata/loaders/xlsx.py
+++ b/visidata/loaders/xlsx.py
@@ -28,10 +28,18 @@ class XlsxIndexSheet(IndexSheet):
 
 
 class XlsxSheet(SequenceSheet):
+    columns = [
+            Column('sog', getter=lambda col, row: 'snot')
+    ]
     def iterload(self):
         worksheet = self.source
         for row in Progress(worksheet.iter_rows(), total=worksheet.max_row or 0):
-            yield list(wrapply(getattr, cell, 'value') for cell in row)
+            yield [{
+                    "value": cell.value,
+                    "fg": cell.fill.fgColor.value if cell.fill else '',
+                    "bg": cell.fill.bgColor.value if cell.fill else ''
+                   } for cell in row]
 
 
 class XlsIndexSheet(IndexSheet):
```

I think I could write a plugin that offers a command to replace each cell value in a column with this dict, perhaps followed automatically by `(`. That way if you have a column where either fgColor or bgColor has semantic value, you can access it easily. Does that seem like the correct approach?

**Reply by saulpw** [reply]
I think the correct approach here is to yield the list of cells directly, and then use AttrColumn of 'value' for the primary value.  This will require unfactoring from the SequenceSheet (which uses ColumnItem), which is a little annoying because we still want the skip/header/sort behavior.  We can expose the fill attributes as hidden columns, I think.  This all probably belongs in the xlsx loader and not a plugin though.

**Reply by hoclun-rigsep** [reply]
def myAddColumn(self, colIndex, colnamelines=None, *args, **kwargs):
        if not colnamelines:
            colnamelines = ['']
        self.addColumn(
            ExprColumn(
                ''.join(map(str, colnamelines)), 
                'row[i].value',
                i=colIndex,
                ),
            ColumnItem(
                ''.join(map(str, colnamelines + ['_cellPyObj'])),
                colIndex,
                width=0
                ),
            ExprColumn(
                ''.join(map(str, colnamelines + ['_fgColor'])), 
                'row[i].fill.fgColor.value',
                i=colIndex,
                width=0
                ),
            ExprColumn(
                ''.join(map(str, colnamelines + ['_bgColor'])), 
                'row[i].fill.bgColor.value',
                i=colIndex,
                width=0
                ),
        )

**Reply by hoclun-rigsep** [reply]
<img width="253" alt="Untitled 2" src="https://user-images.githubusercontent.com/20741358/131738413-537592b4-c8ec-4482-8219-080bd13f3225.png">
<img width="820" alt="Untitled 3" src="https://user-images.githubusercontent.com/20741358/131738414-5ecd231a-a910-404f-82a5-c9f8bad0d026.png">

WIP. I have hidden columns for the cell object itself, the foreground color, and the background color. I guess there is a judgment call here regarding what if any other cell properties to make into columns. Also I imagine we will want to base adding these columns on an option because seeing them all in the columns sheet may be confusing to some?
